src/Query/DateParser.ts

```diff
@@ -75,12 +85,38 @@ export class DateParser {
             const unitOfTime = range === 'week' ? 'isoWeek' : (range as moment.unitOfTime.DurationConstructor);
             dateRange = [dateRange[0].startOf(unitOfTime), dateRange[1].endOf(unitOfTime)];
 
-            // Dates shall be at midnight eg 00:00
-            dateRange.forEach((d) => d.startOf('day'));
+            return DateParser.setDateRangeToStartOfDay(dateRange);
+        }
+
+        return undefined;
+    }
+
+    private static parseSpecificDateRange(input: string): [moment.Moment, moment.Moment] | undefined {
+        const parsingVectors = [
+            { regexp: /[0-9]{4}/, format: 'YYYY', unit: 'year' },
+            { regexp: /[0-9]{4}-Q[1-4]/, format: 'YYYY-Q', unit: 'quarter' },
+            { regexp: /[0-9]{4}-[0-9]{2}/, format: 'YYYY-MM', unit: 'month' },
+            { regexp: /[0-9]{4}-W[0-9]{2}/, format: 'YYYY-WW', unit: 'isoWeek' },
+        ];
 
-            return dateRange;
+        for (const vector of parsingVectors) {
+            const matched = input.match(vector.regexp);
+            if (matched && matched.length === 1 && matched[0] === input) {
+                const range = matched[0];
+                const parsedRange: [moment.Moment, moment.Moment] = [
+                    moment(range, vector.format).startOf(vector.unit as moment.unitOfTime.StartOf),
+                    moment(range, vector.format).endOf(vector.unit as moment.unitOfTime.StartOf),
+                ];
+                return DateParser.setDateRangeToStartOfDay(parsedRange);
+            }
```

Comment:
Nice. I ended up with:

```ts
    private static parseSpecificDateRange(input: string): [moment.Moment, moment.Moment] | undefined {
        const parsingVectors: [RegExp, string, moment.unitOfTime.StartOf][] = [
            [/[0-9]{4}/, 'YYYY', 'year'],
            [/[0-9]{4}-Q[1-4]/, 'YYYY-Q', 'quarter'],
            [/[0-9]{4}-[0-9]{2}/, 'YYYY-MM', 'month'],
            [/[0-9]{4}-W[0-9]{2}/, 'YYYY-WW', 'isoWeek'],
        ];

        for (const [regexp, format, unit] of parsingVectors) {
            const matched = input.match(regexp);
            if (matched && matched.length === 1 && matched[0] === input) {
                const range = matched[0];
                const parsedRange: [moment.Moment, moment.Moment] = [
                    moment(range, format).startOf(unit),
                    moment(range, format).endOf(unit),
                ];
                return DateParser.setDateRangeToStartOfDay(parsedRange);
            }
        }

        return undefined;
    }
```

Reply:
> Nice. I ended up with:
> 
> ```ts
>     private static parseSpecificDateRange(input: string): [moment.Moment, moment.Moment] | undefined {
>         const parsingVectors: [RegExp, string, moment.unitOfTime.StartOf][] = [
>             [/[0-9]{4}/, 'YYYY', 'year'],
>             [/[0-9]{4}-Q[1-4]/, 'YYYY-Q', 'quarter'],
>             [/[0-9]{4}-[0-9]{2}/, 'YYYY-MM', 'month'],
>             [/[0-9]{4}-W[0-9]{2}/, 'YYYY-WW', 'isoWeek'],
>         ];
> 
>         for (const [regexp, format, unit] of parsingVectors) {
>             const matched = input.match(regexp);
>             if (matched && matched.length === 1 && matched[0] === input) {
>                 const range = matched[0];
>                 const parsedRange: [moment.Moment, moment.Moment] = [
>                     moment(range, format).startOf(unit),
>                     moment(range, format).endOf(unit),
>                 ];
>                 return DateParser.setDateRangeToStartOfDay(parsedRange);
>             }
>         }
> 
>         return undefined;
>     }
> ```

This is what I did at first too, but I dislike the loop with `regexp, format, unit`, while  `vector.<something>` makes it crystal clear what is being iterated on. Also the array declaration becomes clearer in this case with `this:..., that:..., andThat:...`. If you prefer the plain array approach let me know.

Reply:
OK no problem, stick with what you've got! ❤️

Reply:
I would be glad to have an argument to change in favor of your implem =)

Reply:
maybe it is better to discuss it live on our next session

Reply:
I'll have one go...

#### Part 1:

I like your earlier version, where you started the types as part of the array...

```ts
        const parsingVectors: [RegExp, string, moment.unitOfTime.StartOf][] = [
            [/[0-9]{4}/, 'YYYY', 'year'],
```

I think that by stating the types at the point where the data is created, they only need to be stated once - and it removes the need for things like ` as moment.unitOfTime.StartOf` which clutter up the implementation...

Put another way, once the language knows the types, we don't have to do work to coerce the types - and our future maintainers don't have to understand why the coercion is needed...

#### Part 2:

And when the loop looks line this:

```ts
        for (const [regexp, format, unit] of parsingVectors) {
```

a whole load of `vector.` strings disappear, which makes the intent stand out rather more clearly, as one doesn't need to read past 1 or two `vector.` to reach the descriptive names...

Reply:
But really I am splitting hairs, as I know that you are keen to explore different techniques - both implementations are definitely good enough.

Reply:
Ok the `as moment.unitOfTime.StartOf` thing is a deal breaker for me =)